Add ability to allow popping visuals out into their own window

andrewkfiedler opened the pull request:
This is the breakdown by commit, which might be easier to review.

- Added popout/cross-window communication to golden layout usage
  - Implemented router and local storage patches for communication and refreshes between windows.

- Remove unused event types

- Add preference syncing between subwindows and main window
  - Implemented preference synchronization between golden-layout subwindows and main window.

- Add default for resultCount to properties
  - Set a default value for resultCount in properties to prevent issues during page loading.

- Fix useSelectionOfLazyResult hook for changing lazyResult
  - Adapted hook to handle changes in lazyResult, especially in subwindows.

- Remove unnecessary conditions around state change callbacks
  - Simplified event emission conditions with state change comparisons.

- Update removeEphemeralState to remove popout dimensions
  - Removed unreliable popout dimensions tracking to avoid unnecessary resizing.

- Fix updates not syncing between windows due to shared objects
  - Deep cloning implemented to prevent updates on shared objects across windows.
  - Updated cloning method for better performance.

- Update geometry for correct updates when lazyResult changes
  - Fixed updates not reflecting due to missing dependencies in useEffects.

- Start looking at stack support in subwindows
  - Introduced disabling to prevent subwindows of subwindows.
  - Refined handling of layout changes in subwindows.

- Add popout tab button and fix histogram update issue
  - Added popout tab button in main windows.
  - Addressed histogram update issue during initialization.

- Add support for subwindows communicating layout changes
  - Enabled subwindows to notify the main window about layout changes without sending a config.

- Add listening for subwindow close events
  - Improved handling of subwindow close events by latching onto windowClose and forwarding the event source.

- Add communication of subwindow navigation events to main window
  - Subwindow navigation blocked, and events sent to main window for proper handling.
  - Updated components to use React Router directly and removed old backbone router compatibility.

- Add handling of subwindows without parents
  - Popup dialog implemented, requesting users to close parentless subwindows.

- Add handling for popups being blocked
  - Addressed scenarios where popups are blocked in a visualization layout with popouts.
  - Users prompted to enable popups and retry or proceed without popups.

- Refactor highlights storage and communication between subwindows
  - Highlights stored in LazyQueryResults and shared between subwindows.
  - Moved transformations to separate functions.
  - Improved typesafety in QueryResponse.
  - Handled highlights effectively between subwindows.

- Add handling for showingResultsForFields and didYouMeanFields
  - Ensured preservation of these fields when navigating between the main window and subwindows.

- Update anchor handling for blob URLs in subwindows
  - Updated handling of blob URLs to prevent forwarding, as they're typically used for downloads.

andrewkfiedler committed:
Adds popout / cross window communication to our golden layout usage

 - In app tsx we add a monkey patch to the router so we can preserve important query params for subwindows from goldenlayout, if they exist.  Without this, windows can go rogue so to speak.
 - We also patch local storage in golden-layout.view tsx so that we can refresh subwindows (very useful for debugging, and likely for users in real life).  Without this, a subwindow from goldenlayout can't be refreshed (it would crash, as it removes it's info on load normally).
 - Refactors a good chunk of the golden-layout.view useEffects into more descriptive hooks, and adds new hooks around setting up cross window communication.

andrewkfiedler committed:
Remove unused event types

andrewkfiedler committed:
Add preference syncing between golden-layout subwindows and main window

andrewkfiedler committed:
Add default for resultCount to properties

 - In some cases the page loads preferences before properties, and this gets set to undefined which can mess up paging.

andrewkfiedler committed:
Fix useSelectionOfLazyResult hook to handle when lazyResult changes

 - This is now possible with subwindows in golden layout, so we need to make sure we update isSelected if the lazyResult changes.

andrewkfiedler committed:
Remove now unnecessary conditions around state change callbacks

 - Since we now do a comparison of what we have when a state change comes in, we no longer need to complicate whether or not to emit events.  Instead the reciever will simply ignore events when they already have the sent information.
 - This fixes the case of the subwindow sending an update around the metacard information itself (such as deleted), which the original window would miss since we were not allowing subwindows to send updates on result information previously (previously we only allowed them to emit selection changes).

andrewkfiedler committed:
Update removeEphemeralState to also remove popout dimensions

 - These appear to be unreliable and constantly change, causing it to appear the user as thought the user has resized windows when they haven't.  For now, we will not track dimensions of popouts.

andrewkfiedler committed:
Fix issue with updates not syncing between windows due to shared objects

 - We do a check for if updates are necessary when windows recieve messages.  Unfortunately, objects are shared across windows, so the window that should update sometimes thinks it's already up to date.  To avoid this, we now deepClone the plain result when taking changes from other windows to avoid accidentally updating their object.  We might need to do the same for status / sorts, but I haven't checked into that yet.
 - Pulls in lodash.cloneDeep as it's benchmarked to be roughly twice as fast as our method of doing JSON.stringify + JSON.parse to clone objects.  I've replaced usages of the old method throughout the code.
 - Also removes generateUUID from Common in favor of delegating to uuid, which we already pull in as a dependency and use in a number of places throughout the codebase.
 - Removes the thumbnail transformation to cache bust from being done in LazyQueryResult (which can cause issues on edit / window syncing, where an update will be missed, or worse we enter infinite loops).  Instead, I've moved this cache bust function to a separate file, and pull it in to be used just in time when folks are using the thumbnail property (right now most places call Common.getImgSrc since the thumbnail is sometimes a url and sometimes a base64 encoding).

andrewkfiedler committed:
Update geometry to correctly update when lazyResult changes

 - We were missing some dependencies in the useEffects, so edits weren't always caught when doing cross window updates.

andrewkfiedler committed:
Start looking at stack support in subwindows

 - Added a bit of disabling to prevent subwindows of subwindows
 - Updated the golden layout state change callback to not run on subwindows, as their config will override the main window config.  Instead, we'll need to communicate layout updates back to the main window, which can then update the openPopouts key of it's config appropriately.

andrewkfiedler committed:
Add a popout tab button, separate from the popout stack button
 - Both only visible in main windows
 - Similar to close tab, popout tab only visible on active tabs
 - Fix an issue with the histogram where update was being called and throwing an error since the histogram wasn't done being initialized yet.

andrewkfiedler committed:
Add support for subwindows communicating layout changes back to the main window

 - We don't need to send a config, just notify the main window that changes are available so it can check if it needs to update.

andrewkfiedler committed:
Move resize trigger to be common between window / subwindow on layout changes

 - This ensures the map view (and anything else dependent on the size of the visuals) will adjust size accordingly as the golden layout layout is adjusted (dragged panels to adjust sizes within the stacks).

andrewkfiedler committed:
Add listening for subwindow close events

 - There's a bit of reordering we have to do and piping of events, as the current golden layout code emits the event for stateChanged before the state is finished changing, and it does not forward the source of the event.  So we latch onto windowClose, and use setTimeout to wait for the state to finish changed (push our request to the end of the stack), and forward the source of the event.

andrewkfiedler committed:
Add communication of subwindow navigation events back to main window (golden layout)

 - Navigation on subwindows is now blocked (aside from user changing it in the top url bar or through the console).
 - Navigation attempts are sent to the main window, which processes them as if the user wants to navigate from there instead.  This prevents a lot of weirdness and is likely the intent of the user.
 - Removes remnants of backwards compatibility for old backbone router, migrating the final two usages of it to use react router directly.
 - Please note the comments on the two new components (for subwindow and window).  Subwindow navigation changes are specific to each react tree of the visuals, so we need to include handling in each visual.  Meanwhile we only need to intercept those in the main window from the single golden layout instance.

andrewkfiedler committed:
Add handling of subwindows without parents
 - For now, popup a dialog asking them to close the parentless window.

andrewkfiedler committed:
Add handling for if popups are blocked

 - If the user has a visualization layout with popouts and popups are blocked, we now handle the scenario.
 - User is prompted to enable popups and retry, or proceed without popups.
 - We can't just retry when they press retry, as the user interacting allows at least one window to open, so instead we attempt to open two and see if it works.

andrewkfiedler committed:
Refactor highlights to keep them stored in LazyQueryResults, add them to the communication between subwindows

 - Updates the highlight handling in LazyQueryResults to be conducive to sharing them between subwindows.
 - Moves transformation to a function and keeps it outside (in QueryResponse)
 - Updates QueryResponse to be slightly more typesafe
 - Highlights now are handled appropriately between subwindows

andrewkfiedler committed:
Add handling for showingResultsForFields and didYouMeanFields

 - Ensure these are preserved when going back and forth between main window and subwindow, otherwise we might lose them

andrewkfiedler committed:
Update anchor handling to allow blob urls to be clicked in subwindows

 - We no longer forward urls that start with blob, as these are typically downloads.

andrewkfiedler committed:
Merge remote-tracking branch 'origin/master' into RE2

jrnorth commented on ui-frontend/packages/catalog-ui-search/src/main/webapp/component/golden-layout/golden-layout.view.tsx:
```diff
@@ -557,23 +653,388 @@ export const GoldenLayoutViewReact = (options: GoldenLayoutViewProps) => {
+              }
+            })
+          )
+        ) {
```
✏️ Recommend something like this instead, to improve readability:
```typescript
const results = Object.values(lazyResults.results).map((lazyResult) => {
  return {
    plain: lazyResult.plain,
    isSelected: lazyResult.isSelected,
  }
})
const callbackResults = Object.values(eventData.lazyResults.results).map((lazyResult) => {
  return {
    plain: lazyResult.plain,
    isSelected: lazyResult.isSelected,
  }
})
if (!_.isEqual(results, callbackResults)) {
```

jrnorth commented on ui-frontend/packages/catalog-ui-search/src/main/webapp/component/golden-layout/golden-layout.view.tsx:
```diff
@@ -557,25 +861,502 @@ export const GoldenLayoutViewReact = (options: GoldenLayoutViewProps) => {
+
   return (
     <div data-element="golden-layout" className="is-minimised h-full w-full">
+      {popupHandlingState === 'blocked' ? (
```
✏️ `{popupHandlingState === 'blocked' && (`

andrewkfiedler replied:
I've never really liked doing the && style of conditional rendering.  I didn't really have a good reason why, other than it wasn't obvious to me.  So I googled to see if there was a more comprehensive discussion on it.

I think either are okay, but it looks like folks say there is a possibility of messing up with the && if we aren't careful around falsey things (these cases in the code are explicit though, so it doesn't apply).

https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx

I think for me it's more that I like how clear it is to see it's a conditional render, and later on if we wanted we could return something else instead of null.  That and I'd be worried that as we add on to a && instance later on, we might mess up with the falsey stuff.

jrnorth replied:
Good article. I might start writing it that way, too.

jrnorth commented on ui-frontend/packages/catalog-ui-search/src/main/webapp/component/golden-layout/golden-layout.view.tsx:
```diff
@@ -135,6 +158,19 @@ const GoldenLayoutComponentHeader = ({
         </Grid>
+        <Grid item>
+          {!tab.contentItem.layoutManager.isSubWindow &&
+          tab.closeElement[0].style.display !== 'none' ? (
```
✏️ Could use another `&&` instead of a ternary.

jrnorth commented on ui-frontend/packages/catalog-ui-search/src/main/webapp/component/golden-layout/golden-layout.view.tsx:
```diff
@@ -454,6 +671,19 @@ const GoldenLayoutToolbar = ({ stack }: { stack: any }) => {
                 <AllOutIcon />
               </Button>
             </Grid>
+            {stack.layoutManager.isSubWindow ? null : (
```
✏️ `{!stack.layoutManager.isSubWindow && (`

jrnorth commented on ui-frontend/packages/catalog-ui-search/src/main/webapp/component/golden-layout/golden-layout.view.tsx:
```diff
@@ -135,6 +158,19 @@ const GoldenLayoutComponentHeader = ({
+          tab.closeElement[0].style.display !== 'none' ? (
+            <Button
+              data-id="popout-tab-button"
+              onClick={() => {
+                tab.contentItem.popout()
+              }}
```
✏️ `onClick={tab.contentItem.popout}`

andrewkfiedler replied:
So one of the reasons I don't do this, is a lot of old libraries tend to depend on `this` binding, and this would muck with that.  I double checked to see if that was the case here, and it indeed does depend on on this, throwing an error if we try to set it as a direct call.

So in general I tend to not set onClick to function calls directly (unless it's our custom code, but even then, I'd rather not think, despite modern js not relying as much on implicit this binding), just so I don't have to think about it.

jrnorth replied:
Good point. I didn't know that.

jrnorth commented on ui-frontend/packages/catalog-ui-search/src/main/webapp/component/golden-layout/golden-layout.view.tsx:
```diff
@@ -151,11 +187,165 @@ const GoldenLayoutComponentHeader = ({
+        <DialogActions>
+          <Button
+            variant="contained"
+            onClick={() => {
```
✏️ `onClick={window.close}`

jrnorth commented on ui-frontend/packages/catalog-ui-search/src/main/webapp/component/golden-layout/golden-layout.view.tsx:
```diff
@@ -454,6 +671,19 @@ const GoldenLayoutToolbar = ({ stack }: { stack: any }) => {
+              <Grid item>
+                <Button
+                  data-id="popout-layout-button"
+                  onClick={() => {
+                    stack.popout()
+                  }}
```
✏️ `onClick={stack.popout}`

jrnorth commented on ui-frontend/packages/catalog-ui-search/src/main/webapp/js/cach-bust-url.tsx:
Filename should be `cache-bust-url.tsx`

jrnorth reviewed (approved):
Great work, all the commits and code are very well-commented.

andrewkfiedler committed:
Update comparison in useConsumeState change to use isEqualWith

 - This should be faster compared to stringify, and easier to understand for devs.

andrewkfiedler committed:
Fix spelling of cache

andrewkfiedler committed:
Merge remote-tracking branch 'origin/master' into RE2

alexabird commented:
build now

andrewkfiedler committed:
Update test to fit new cache bust behavior

alexabird commented:
build now
